include/rmm/mr/host/pinned_memory_resource.hpp

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "host_memory_resource.hpp"
+
+#include <cuda_runtime_api.h>
+#include <cstddef>
+#include <utility>
+
+namespace rmm {
+namespace mr {
+
+/**---------------------------------------------------------------------------*
+ * @brief A `host_memory_resource` that uses `cudaMallocHost` to allocate
+ * pinned/page-locked host memory.
+ *
+ * See https://devblogs.nvidia.com/how-optimize-data-transfers-cuda-cc/
+ *---------------------------------------------------------------------------**/
+class pinned_memory_resource final : public host_memory_resource {
+ private:
+  /**---------------------------------------------------------------------------*
+   * @brief Allocates pinned memory on the host of size at least `bytes` bytes.
+   *
+   * The returned storage is aligned to the specified `alignment` if supported,
+   * and to `alignof(std::max_align_t)` otherwise.
+   *
+   * @throws std::bad_alloc When the requested `bytes` and `alignment` cannot be
+   * allocated.
+   *
+   * @param bytes The size of the allocation
+   * @param alignment Alignment of the allocation
+   * @return void* Pointer to the newly allocated memory
+   *---------------------------------------------------------------------------**/
+  void *do_allocate(std::size_t bytes, std::size_t alignment =
+                                           alignof(std::max_align_t)) override {
+    // don't allocate anything if the user requested zero bytes
+    if (0 == bytes) {
+      return nullptr;
+    }
+    // allocate memory for bytes, plus potential alignment correction,
+    // plus store of the correction offset
+    void *p{nullptr};
+    auto status = cudaMallocHost(&p, (bytes + alignment + sizeof(std::size_t)));
+    if (cudaSuccess != status) {
+      throw std::bad_alloc{};
+    }
+
+    std::size_t ptr_int = reinterpret_cast<std::size_t>(p);
```

Comment:
This logic is used in two different memory resource implementations. Perhaps factor it out for reuse?

Reply:
Yeah, I was just starting to work on that now :)

Reply:
Done. See `aligned_allocate/deallocate`.